File: Python/766. ToeplitzMatrix.py
#A matrix is Toeplitz if every diagonal from top-left to bottom-right has the same element.
#Now given an M x N matrix, return True if and only if the matrix is Toeplitz.

class Solution:
    def isToeplitzMatrix(self, matrix: List[List[int]]) -> bool:
        c = True
        rows = len(matrix)
        cols = len(matrix[0])
        
        for x in range(rows):
            for y in range(cols):
                if x != 0 and y != 0: 
                    if matrix[x][y] != matrix[x-1][y-1]:
                        c = False
                        break
        if c == False:
            return False
        else:
            return True 

File: C++/766. Toeplitz Matrix.cpp
// problem - 766. Toeplitz Matrix

/*
    Given an m x n matrix, return true if the matrix is Toeplitz. Otherwise, return false.

    A matrix is Toeplitz if every diagonal from top-left to bottom-right has the same elements.
*/

class Solution {
public:
    bool isToeplitzMatrix(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        
        if(m == 1 || n == 1) return true;
        
        // traversing diagonally
        for(int i=1; i<n; i++) {
            for(int j=1; j<m; j++) {
               if(matrix[i][j] != matrix[i-1][j-1]) return false;
            }
        }
        
        return true;
    }
};

// Time Complexity = O(m*n) [where m is row length and n is column length]
// Space Complexity = O(1)